### ISSUE TITLE
Flutter Web: ListView lags a lot on mobile browsers, inconsistent performance on different devices on different browsers

### ISSUE THREAD

**Issue by Auties00** [flutter/flutter#59332]
1. A listview with "complex" widgets inside doesn't lag on Chrome or Firefox for Windows or 
    MacOS, but it's completely unusable on chrome for Android and IOS

2.  Here is an example, but probably any widget will produce the same effect:
    Important Part: https://gist.github.com/Auties00/63dab901d0f203a619481398db3d1472
    Website: https://auties00.github.io/#/
    Source code: https://github.com/Auties00/hytaleserverhosting
3.  As far as I know, I cannot debug a flutter web project on a mobile browser, so this data cannot 
     be provided

4. Target Platform: Web
    Browser: Any mobile browser for android or ios, I tried chrome, safari and firefox
    Devices: Huawei p30 pro, Huawei p30 lite, iPhone 11 Pro Max, iPhone XR

5. Flutter Doctor
```
[√] Flutter (Channel master, 1.19.0-5.0.pre, on Microsoft Windows [Versione 10.0.18363.900], locale it-IT)
    • Flutter version 1.19.0-5.0.pre at C:\flutter
    • Framework revision 28138ec5e3 (8 days ago), 2020-06-04 00:33:01 -0400
    • Engine revision 2608f2ee9f
    • Dart version 2.9.0 (build 2.9.0-13.0.dev 54adfeb93f)

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.4.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.4.29709.97

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] VS Code (version 1.41.1)
    • VS Code at C:\Users\Ale\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.7.1

[√] Connected device (3 available)
    • Windows    • Windows    • windows-x64    • Microsoft Windows [Versione 10.0.18363.900]
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 83.0.4103.97
```

**Comment by Auties00** [comment]
Update:
The issue is fixed on:
Samsung galaxy note 9
Huawei p20 lite

The issue remains for other devices, but it's much better:
https://gist.github.com/Auties00/e915e80e8db32364466ef6ddccc9eed1

**Comment by TahaTesser** [comment]
Hi @Auties00 
Can you please provide and a complete reproducible runnable minimal code sample without using third party packages
Thank you

**Comment by Auties00** [comment]
Any and all widgets will produce the same result. It's as simple as creating a new flutter project, setting as the body of a scaffold a LIstView and inserting some containers as children

**Comment by TahaTesser** [comment]
Hi @Auties00 
Tried code sample based your description, no issues on mobile web

https://elastic-leakey-a57280.netlify.app/#/

<details> 
<summary>Code Sample</summary> 

``` 
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      body: ListView(
        children: [
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
          Container(
            color: Colors.blue,
            child: Image.network('https://i.imgur.com/QCNbOAo.png'),
          ),
          SizedBox(height: 10),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () {},
      ),
    );
  }
}

``` 

</details> 

<details> 
<summary>flutter doctor -v</summary> 

``` 
[✓] Flutter (Channel master, 1.20.0-1.0.pre.91, on Mac OS X 10.15.5 19F101,
    locale en-GB)
    • Flutter version 1.20.0-1.0.pre.91 at /Users/tahatesser/Code/flutter_master
    • Framework revision 9c3f0faa6d (10 hours ago), 2020-06-17 18:47:54 -0400
    • Engine revision 965fbbed17
    • Dart version 2.9.0 (build 2.9.0-14.0.dev 2b917f5b6a)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.0
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.46.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (5 available)
    • SM M305F      • 32003c30dc19668f          • android-arm64  • Android 10
      (API 29)
    • Taha’s iPhone • 00008020-001059882212002E • ios            • iOS 13.5.1
    • macOS         • macOS                     • darwin-x64     • Mac OS X
      10.15.5 19F101
    • Web Server    • web-server                • web-javascript • Flutter Tools
    • Chrome        • chrome                    • web-javascript • Google Chrome
      83.0.4103.106

• No issues found!
``` 

</details>

**Comment by Auties00** [comment]
Just tried it on chrome mobile and it works perfectly, on Firefox mobile it's very laggy. Also I was able to fix the issue on Chrome and Safari, now the scrolling is really smooth but on on Firefox Mobile every flutter site still lags and on Pixel devices you have to hold the screen to scroll

**Comment by Auties00** [comment]
So this got stranger. On my huawei p30 lite the website works perfectly on chrome, but not on firefox. On the Pixel 3, it performs better on firefox and on chrome scrolling is bugged. How is this even possible?

**Comment by Auties00** [comment]
So what now?

**Comment by yjbanov** [comment]
What's the performance like if you remove the platform view (e.g. replace it with an image)? Curious if that's the root cause.

**Comment by Auties00** [comment]
> What's the performance like if you remove the platform view (e.g. replace it with an image)? Curious if that's the root cause.

Same thing

**Comment by Auties00** [comment]
It looks like SKIA is the fix to this issue, enabling it fixed the issue, But for some reason I can't scroll on some devices when Skia is enabled

**Comment by rajveermalviya** [comment]
+1

But not just Firefox Mobile, on Firefox desktop it's slow/laggy scrolling when compared to Chrome.
Also this behavior is same on both rendering back-ends (html/canvaskit).

**Comment by Auties00** [comment]
> +1
> 
> But not just Firefox Mobile, on Firefox desktop it's slow/laggy scrolling when compared to Chrome.
> Also this behavior is same on both rendering back-ends (html/canvaskit).

Let's say it makes things better, still very laggy though

**Comment by Ajmaljalal** [comment]
have the same issue on mobile safari and chrome browsers. it is supper lagy and slow

**Comment by Lock128** [comment]
I've gotten the same feedback, that scrolling is unusable for Firefox Mobile 

> flutter doctor -v
> [√] Flutter (Channel stable, 2.0.1, on Microsoft Windows [Version 10.0.18363.1379], locale en-US)
>     • Flutter version 2.0.1 at C:\ToolsJK\flutter
>     • Framework revision c5a4b4029c (11 days ago), 2021-03-04 09:47:48 -0800
>     • Engine revision 40441def69
>     • Dart version 2.12.0

**Comment by xenSlayer** [comment]
same issue. The performance is too bad. It's lags like hell in ios safari but works smooth in laptop chrome.

**Comment by TimLampen** [comment]
I was able to fix this by wrapping my ListView using a Scaffold.

**Comment by MahmoodBakhshayesh** [comment]
@pedromassangocode  This problem still exist in phone browsers even in flutter 2.0.5 
simple item in lists are fine but if you add some complex widget like Animated Container  or ExpansionTile the list and animation will get laggy 
unfortunately this keeps my company from switching completely for every platform to flutter

**Comment by kraa** [comment]
Flutter 3.3.6I have the same problem.  On some devices scroll work smooth , but on some it's terrible (Huawei p10) no matter what browser used..

**Comment by omar-hanafy** [comment]
Same here, even with a simple app. Check this out.
 Website: omarkhaled.web.app
 Browser: Safari
 Device(OS): iPhone 13 (iOS 16.3.1)
I can see the frames drops to like 20 frame and no smooth scrolling experience as well.

**Comment by exaby73** [comment]
I cannot reproduce this issue on my iOS device (12 Pro Max), with Chrome, Firefox or Safari. @Auties00 can you confirm on this?

**Comment by RoarGronmo** [comment]
Still a problem in ListView.builder when 
Using card
+ You have Icons
+ You have wrapped some elements in `IntrinsicHeight` to "redraw" to catch height i.e. for VerticalDivider.

Tested in real device Samsung S9+ SM-G965F, One UI v2.5, Android 10, Chrome version 112.0.5615.136

Flutter version
`Flutter (Channel beta, 3.10.0-1.4.pre, on Microsoft Windows [Version 10.0.22621.1555], locale nb-NO)`

**Comment by Auties00** [comment]
> I cannot reproduce this issue on my iOS device (12 Pro Max), with Chrome, Firefox or Safari. @Auties00 can you confirm on this? 

It's still an issue on mobile. I found that attaching an HTML scrollable div to the  Flutter Scrollable fixes the issue on firefox. On chrome I couldn't reproduce the issue, but it could be because my phone is a OnePlus 8T which is still a modern phone. I don't have the technical insight to know, but I don't think this issue can be solved on older devices without using the browser's default scroll behaviour instead of the transformation system that flutter uses on the canvas.

**Comment by iJack93** [comment]
This keeps happening even on the material demo 3 online. Here my videos: https://github.com/flutter/flutter/issues/119322#issuecomment-1683474308
This behavior happend on Safari Mobile, Chrome Mobile and Firefox mobile

**Comment by AleXoTroN** [comment]
I also experience this and another user that I know also has that problem at his project.

If I open his web app from a mobile browser on my pixel 5, then scrolling is extremely laggy:
https://e1547.clynamic.net/

**Comment by Ultranmus** [comment]
Same issue in my portfolio website ,sometime the scroll is lagging in chrome in mobile device. But is laptop it works totally fine.

**Comment by AleXoTroN** [comment]
> Same issue in my portfolio website ,sometime the scroll is lagging in chrome in mobile device. But is laptop it works totally fine.
> 

I actually figured out how to solve it:
Just force flutter to use the canvaskit web-renderer when building your web app. It's completely smooth for me now on mobile

**Comment by Ultranmus** [comment]
@AleXoTroN,

I'm interested in trying your solution, but I've encountered an issue. When I attempt to force Flutter to use the canvaskit web-renderer for building a web app, I encounter errors with some Network images that don't occur when using the html web-renderer. Could you please provide guidance on how to address this issue?

**Comment by AleXoTroN** [comment]
I just did "flutter build web --release --web-renderer=canvaskit" and everything works fine. You should maybe check the errors in the web console regarding the images when you open your web page from your computer

**Comment by Ultranmus** [comment]
@AleXoTroN 
The issue is not in image it is in flutter itself, for canvaskit it throw error but for html web renderer it work fine.
You can check out the image resource's i am using - https://pngimg.com/uploads/mysql/small/mysql_PNG36.png
There are many more image that are not processed in canvaskit web renderer but work fine in html. I tried varies network image packages like cached_network_image but that also do not work for such image in canvaskit web renderer.

**Comment by Boahen123** [comment]
Seems this issue has still not been fixed. I'm looking forward to a fix. In the meantime, can anyone help out with a workaround, my portfolio site is built with flutter web and scrolling lags on chrome browser with Samsung Galaxy A35

**Comment by AleXoTroN** [comment]
> Seems this issue has still not been fixed. I'm looking forward to a fix. In the meantime, can anyone help out with a workaround, my portfolio site is built with flutter web and scrolling lags on chrome browser with Samsung Galaxy A35

I actually solved it by forcing flutter to use canvaskit. Now I have 60-90fps.

**Comment by Ultranmus** [comment]
@AleXoTroN 
But if we canvaskit  then all type of network image are not supported on flutter.

**Comment by Boahen123** [comment]
> > Seems this issue has still not been fixed. I'm looking forward to a fix. In the meantime, can anyone help out with a workaround, my portfolio site is built with flutter web and scrolling lags on chrome browser with Samsung Galaxy A35
> 
> I actually solved it by forcing flutter to use canvaskit. Now I have 60-90fps.

Thank you for this. Kindly guide me on how to do this exactly. thanks

**Comment by knottx** [comment]
Still issue for 4 years 😞

**Comment by Kataglyphis** [comment]
> Still issue for 4 years 😞

I can relate ... Still the same issue.

**Comment by AleXoTroN** [comment]
> > Still issue for 4 years 😞
> 
> I can relate ... Still the same issue.

It's gone for me since 2 years

**Comment by AleXoTroN** [comment]
> > Still issue for 4 years 😞
> 
> I can relate ... Still the same issue.

Here is the solution:
https://github.com/flutter/flutter/issues/59332#issuecomment-1781692325

**Comment by huycozy** [comment]
I can reproduce this issue on Firefox browser (on both Android and macOS desktop) using the [given sample code](https://github.com/Auties00/hytaleserverhosting). The issue seems to appear on Flutter `stable` channel 3.19.5 but it seems to be fixed on the latest `master` channel `3.22.0-11.0.pre.14`.

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.19.5, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.19.5 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 300451adae (8 days ago), 2024-03-27 21:54:07 -0500
    • Engine revision e76c956498
    • Dart version 3.3.3
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.87.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 123.0.6312.106

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.22.0-11.0.pre.14, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.22.0-11.0.pre.14 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 817630ccf1 (51 minutes ago), 2024-04-15 22:56:25 -0400
    • Engine revision 76fbe31a96
    • Dart version 3.5.0 (build 3.5.0-61.0.dev)
    • DevTools version 2.35.0-dev.8
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.88.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.86.0

[✓] Connected device (4 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D                         • android-arm64  • Android 11 (API 30)
    • iPhone (mobile)  • d9a94afe2b649fef56ba0bfeb052f0f2a7dae95e • ios            • iOS 15.8 19H370
    • macOS (desktop)  • macos                                    • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome                                   • web-javascript • Google Chrome 123.0.6312.124

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

@Auties00 (and also others facing the issue) Please try running app again on the latest master channel and confirm if this issue is fixed on your end. Thanks!
